### PR TITLE
Add grid search algorithm

### LINKED ISSUE #508: Add basic Grid Search

### PR DESCRIPTION
**Pull request by bouthilx**
[Fixes #508]

### CODE REVIEW

File: src/orion/algo/gridsearch.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+
+        while n_trials > max_trials:
+            coordinates = []
+            capped_values = []
+            for name, dim in space.items():
+                capped_value = max(n_values[name] - adjust, 1)
+                capped_values.append(capped_value)
+                coordinates.append(list(grid(dim, capped_value)))
+
+            if all(value <= 1 for value in capped_values):
```
Comment (donglinjy) [inline src/orion/algo/gridsearch.py:159]: looks like capped_value for all dimension will always be the same at this place.  So we want `raise` if any dimension `n_values` is adjusted to 1 or all the dimension is adjust?

File: tests/unittests/algo/test_gridsearch.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    space.register(dim1)
+    space.register(dim2)
+    space.register(dim3)
+    space.register(dim4)
+
+    with pytest.raises(ValueError) as exc:
+        GridSearch.build_grid(space, {k: 5 for k in space}, 10)
+
+    assert exc.match(
+        "Cannot build a grid smaller than 10. "
```
Comment (donglinjy) [inline tests/unittests/algo/test_gridsearch.py:137]: Same as above, for the space, there could be a grid with size `8`, where first 2 dimension with only one candidate. Is it valid?

File: src/orion/algo/gridsearch.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+    if dim.prior_name.endswith("reciprocal"):
+        a, b = dim.interval()
+        return list(_log_grid(a, b, num))
+    elif dim.prior_name.endswith("uniform"):
+        a, b = dim.interval()
+        return list(_lin_grid(a, b, num))
+    else:
+        raise TypeError(
+            "Grid Search only supports `loguniform`, `uniform` and `choices`: "
+            "`{}`".format(dim.prior_name)
```
Comment (donglinjy) [inline src/orion/algo/gridsearch.py:90]: Seems same error message as the one inside the `def grid(dim, num):`

File: src/orion/algo/gridsearch.py (hunk 1 above)
Comment (bouthilx) [reply src/orion/algo/gridsearch.py:159]: `n_values[name]` can vary from one dimension to another. We want to keep adjusting until grid is >= max_trials or until there is no more adjustment possible (all dims have only 1 point each). If this is still larger than max_trials, raise ValueError.

Comment (bouthilx) [reply src/orion/algo/gridsearch.py:159]: `n_values[name]` can vary from one dimension to another. We want to keep adjusting until grid is >= max_trials or until there is no more adjustment possible (all dims have only 1 point each). If this is still larger than max_trials, raise ValueError.

File: tests/unittests/algo/test_gridsearch.py (hunk 2 above)
Comment (bouthilx) [reply tests/unittests/algo/test_gridsearch.py:137]: The source of the (intentional) problem is the 2 categorical dimensions with 5 choices each. The grid always picks all choices, no matter what `n_values` is. This leads to a smallest possible grid of 1 * 1 * 5 * 5 > 10.

File: src/orion/algo/gridsearch.py (hunk 3 above)
Comment (bouthilx) [reply src/orion/algo/gridsearch.py:90]: Oh indeed, I should change the error message for grid(). Thanks!

File: tests/unittests/algo/test_gridsearch.py (hunk 2 above)
Comment (donglinjy) [reply tests/unittests/algo/test_gridsearch.py:160]: oh, I made a mistake, thought `Categorical("dim3", "abcde")` had 2 values, but it had 5.

File: src/orion/algo/gridsearch.py (hunk 1 above)
Comment (donglinjy) [reply src/orion/algo/gridsearch.py:161]: How about a space like below. if `max_trials` is 10, when `n_values` = 2 for all dims, the grid size is 16, but in a middle of adjustment such as `n_values[dim1]=1, n_values[dim2]=2, n_values[dim3]=2, n_values[dim4]=2`, the grid size is 8, which looks like a valid gride.

```
    dim1 = Real("dim1", "uniform", 0, 1)
    dim2 = Integer("dim2", "uniform", 0, 10)
    dim3 = Categorical("dim3", "ab")
    dim4 = Categorical("dim4", "cd")
```

Comment (bouthilx) [reply src/orion/algo/gridsearch.py:161]: Yeah we could do the adjustment in a more efficient way, adjusting by 1 for all incrementally is very basic. Lazy me is tempted to say current version is good enough however. :) Do you think leaving this improvement for later would be harmful for the usability of the algorithm? 

Or perhaps I could just add an `if` test in the for-loop to break earlier if we reach n_trials < max_trials.

Comment (donglinjy) [reply src/orion/algo/gridsearch.py:161]: I think current implementation is okay, just want to confirm this is what we want to do. :)

**Review by donglinjy — APPROVED** [review]

**Pull request merged**